pkg/virt-controller/watch/drain/evacuation/evacuation.go

```diff
@@ -371,18 +371,27 @@ func (c *EvacuationController) sync(node *k8sv1.Node, vmisOnNode []*virtv1.Virtu
 		Effect: k8sv1.TaintEffectNoSchedule,
 	}
 
+	numOfActiveMigrations := len(activeMigrations)
+
 	vmisToMigrate := vmisToMigrate(node, vmisOnNode, taint)
 	if len(vmisToMigrate) == 0 {
 		return nil
 	}
 
 	migrationCandidates, nonMigrateable := c.filterRunningNonMigratingVMIs(vmisToMigrate, activeMigrations)
-	activeMigrationsFromThisSourceNode := c.numOfVMIMForThisSourceNode(vmisOnNode, activeMigrations)
+	numOfActiveMigrationsFromThisSourceNode := c.numOfVMIMForThisSourceNode(vmisOnNode, activeMigrations)
 	maxParallelMigrationsPerOutboundNode :=
 		int(*c.clusterConfig.GetMigrationConfiguration().ParallelOutboundMigrationsPerNode)
 	maxParallelMigrations := int(*c.clusterConfig.GetMigrationConfiguration().ParallelMigrationsPerCluster)
-	freeSpotsPerCluster := maxParallelMigrations - len(activeMigrations)
-	freeSpotsPerThisSourceNode := maxParallelMigrationsPerOutboundNode - activeMigrationsFromThisSourceNode
+
+	if (numOfActiveMigrations >= maxParallelMigrations) ||
```

Comment:
+1
TBH I'm also confused :)

Reply:
@xpivarc @iholder101 It can happen that `freeSpots` is negative, and also `len(migrationCandidates) ==0 && len(nonMigrateable) == 0`, its when we have VMIM created for all the VMIs, Migrations are pending, and we reduce `maxParallelMigrations` from higher value to something very small

Reply:
I will attach the panic soon, I think I didn't capture it, will have to reproduce

Reply:
In order to make a lot of migrations pending, I've used `nodeSelector` in VMI, and labelled only one node. Then triggered the drain, then reduced `maxParallelMigrations`

Reply:
Hey @enp0s3.
I'm trying to understand what's wrong with the scenario you just outlined.

> @xpivarc @iholder101 It can happen that `freeSpots` is negative, and also `len(migrationCandidates) ==0 && len(nonMigrateable) == 0`

If  `len(migrationCandidates) ==0 && len(nonMigrateable) == 0`, this means there are no VMIs to evacuate, therefore there is no reason to wake up again in the future.

> its when we have VMIM created for all the VMIs, Migrations are pending, and we reduce `maxParallelMigrations` from higher value to something very small

If `freeSpotsPerCluster` is indeed negative, this means that `freeSpots` would be less than zero, which means we would return nil.

Am I missing something here?

Reply:
>@xpivarc @iholder101 It can happen that freeSpots is negative, and also len(migrationCandidates) ==0 && len(nonMigrateable) == 0, 

Ok, this is critical. Imho we should return when there are no free spots! But we continue, is this what you are trying to say?

>its when we have VMIM created for all the VMIs, Migrations are pending, and we reduce maxParallelMigrations from higher value to something very small

Messing with `maxParallelMigrations` should not mess with our logic as long as we check it only once ;)

Reply:
@xpivarc Yes you are right, negative means we don't have free spots and nevertheless, we continue!

Reply:
So IMHO the only change needed is to fix the `return`. Still, I will wait for your reproducer (and preferably unit test) to double-check this.

Reply:
Igor and I talked offline. I agree with @xpivarc that fixing the return would be enough, but I think the following would be even better.

Firstly, this should be executed right after getting `migrationCandidates` and `nonMigrateable`:
```go
migrationCandidates, nonMigrateable := c.filterRunningNonMigratingVMIs(vmisToMigrate, activeMigrations)
if len(migrationCandidates) == 0 && len(nonMigrateable) == 0 {
	return nil
}
```

Then, change the `freeSpots <= 0` condition to the following:
```go
if freeSpots <= 0 {
	c.Queue.AddAfter(node.Name, 5*time.Second)
	return nil
}
```

With this the code is not only clearer, but safer and more performant as we return as soon as we are certain we have nothing left to do

Reply:
Works for me either way.